src/gadgets/curves/bn256/ec_pairing.rs

```diff
@@ -0,0 +1,303 @@
+    CS: ConstraintSystem<F>,
+{
+    resultant_f: BN256Fq12NNField<F>,
+    _marker: std::marker::PhantomData<CS>,
+}
+
+impl<F, CS> FinalExpEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    /// This function computes the final exponentiation for the BN256 curve.
+    pub fn evaluate(cs: &mut CS, f: &mut BN256Fq12NNField<F>) -> Self {
+        let mut easy_part_f = Self::easy_part(cs, f);
+        let hard_part = Self::hard_part(cs, &mut easy_part_f);
+        Self {
+            resultant_f: hard_part,
+            _marker: std::marker::PhantomData::<CS>,
+        }
+    }
+
+    /// This function computes the easy part of the final exponentiation, that is
+    /// for given f \in `F_{p^{12}}` it computes `f^{(p^6-1)(p^2+1)}`.
+    pub fn easy_part(cs: &mut CS, f: &mut BN256Fq12NNField<F>) -> BN256Fq12NNField<F> {
+        // f1 <- f^(p^6 - 1)
+        let mut f1 = f.inverse(cs);
+        let mut fp6 = f.frobenius_map(cs, 6);
+        let mut f1 = f1.mul(cs, &mut fp6);
+
+        // f2 <- f1^(p^2 + 1)
+        let mut fp2 = f1.frobenius_map(cs, 2);
+        let f2 = f1.mul(cs, &mut fp2);
+
+        f2
+    }
+
+    /// Computes the hard part of the final exponentiation using method by Aranha et al.
+    /// For details, see https://eprint.iacr.org/2016/130.pdf, _Algorithm 2_.
+    pub fn hard_part(cs: &mut CS, f: &mut BN256Fq12NNField<F>) -> BN256Fq12NNField<F> {
+        let u = BN256Fq::from_str(CURVE_PARAMETER).unwrap();
+        let u_div_2 = BN256Fq::from_str(CURVE_DIV_2_PARAMETER).unwrap();
+
+        // 1. t0 <- f^2; 2. t1 <- t0^u; 3. t2 <- t1^(u/2);
+        // 4. t3 <- f^(-1); 5. t1 <- t3*t1.
+        let mut t0 = f.square(cs);
+        let mut t1 = t0.pow(cs, u);
+        let mut t2 = t1.pow(cs, u_div_2);
+        let mut t3 = f.inverse(cs);
+        let mut t1 = t3.mul(cs, &mut t1);
+
+        // 6. t1 <- t1^{-1}; 7. t1 <- t1*t2
+        let mut t1 = t1.inverse(cs);
+        let mut t1 = t1.mul(cs, &mut t2);
+
+        // 8. t2 <- t1^u
+        let mut t2 = t1.pow(cs, u);
+
+        // 9. t3 <- t2^u; 10. t1 <- t1^{-1}; 11. t3 <- t1*t3
+        let mut t3 = t2.pow(cs, u);
+        let mut t1 = t1.inverse(cs);
+        let mut t3 = t1.mul(cs, &mut t3);
+
+        // 12. t1 <- t1^{-1}; 13. t1 <- t1^{p^3}; 14. t2 <- t2^{p^2};
+        // 15. t1 <- t1*t2
+        let mut t1 = t1.inverse(cs);
+        let mut t1 = t1.frobenius_map(cs, 3);
+        let mut t2 = t2.frobenius_map(cs, 2);
+        let mut t1 = t1.mul(cs, &mut t2);
+
+        // 16. t2 <- t3^u; 17. t2 <- t2*t0; 18. t2 <- t2*f
+        let mut t2 = t3.pow(cs, u);
+        let mut t2 = t2.mul(cs, &mut t0);
+        let mut t2 = t2.mul(cs, f);
+
+        // 19. t1 <- t1*t2; 20. t2 <- t3^p; 21. t1 <- t1*t2
+        let mut t1 = t1.mul(cs, &mut t2);
+        let mut t2 = t3.frobenius_map(cs, 1);
+        let t1 = t1.mul(cs, &mut t2);
+
+        t1
+    }
+}
+
+/// This function computes the pairing function for the BN256 curve.
+pub fn ec_mul<F, CS>(
```

Comment:
nit: i think you want to name this `ec_pairing` :)

Reply:
Haha, yeah, thank you for noticing :)